File: medium/KClosestPoints/Javascript/solution1.js
/**
 * @param {number[][]} points
 * @param {number} K
 * @return {number[][]}
 */
// Time: O(nlog(n)) | Space: O(n)
var kClosest = function(points, K) {
    if (points.length === K) {
        return points
    }
  return  points.sort((a,b) => distance(a) - distance(b)).slice(0,K)
};

const distance = (point) => {
        return (point[0] ** 2) + (point[1] ** 2)
        
    }

File: medium/KClosestPoints/Java/solution1.java
class Solution {
    // Time: O(nlog(n)) | Space: O(n)
    public int[][] kClosest(int[][] points, int K) {
        // If there are only K number of points those the closest points
        if(K == points.length) return points; 
        // Using priority queue to add points in and order them by distance form origin
        PriorityQueue<int[]> maxHeap = new PriorityQueue<>((a,b) -> (b[0]*b[0] + b[1]*b[1] - (a[0]*a[0] + a[1]*a[1])));
        for(int[] point: points) {
            maxHeap.add(point);
            if(maxHeap.size() > K) 
                maxHeap.remove();
        }
        return maxHeap.toArray(new int[0][0]);
    }
}
